x-pack/plugins/security_solution/public/detection_engine/rule_management/components/rule_details/json_diff/mark_edits.tsx

```diff
@@ -0,0 +1,214 @@
+
+      switch (type) {
+        case DmpChangeType.INSERT:
+          newDiffs.push(diff);
+          break;
+        case DmpChangeType.DELETE:
+          oldDiffs.push(diff);
+          break;
+        default:
+          oldDiffs.push(diff);
+          newDiffs.push(diff);
+          break;
+      }
+
+      return [oldDiffs, newDiffs];
+    },
+    [[], []]
+  );
+}
+
+function splitDiffToLines(diffs: Diff[]): Diff[][] {
+  return diffs.reduce<Diff[][]>(
+    (lines, [type, value]) => {
+      const currentLines = value.split('\n');
+
+      const [currentLineRemaining, ...nextLines] = currentLines.map(
+        (line: string): Diff => [type, line]
+      );
+      const next: Diff[][] = [
+        ...lines.slice(0, -1),
+        [...lines[lines.length - 1], currentLineRemaining],
+        ...nextLines.map((line) => [line]),
+      ];
+      return next;
+    },
+    [[]]
+  );
+}
+
+function diffsToEdits(diffs: Diff[], lineNumber: number): RangeTokenNode[] {
+  const output = diffs.reduce<[RangeTokenNode[], number]>(
+    // eslint-disable-next-line @typescript-eslint/no-shadow
+    (output, diff) => {
+      const [edits, start] = output;
+      const [type, value] = diff;
+      if (type !== DmpChangeType.EQUAL) {
+        const edit: RangeTokenNode = {
+          type: 'edit',
+          lineNumber,
+          start,
+          length: value.length,
+        };
+        edits.push(edit);
+      }
+
+      return [edits, start + value.length];
+    },
+    [[], 0]
+  );
+
+  return output[0];
+}
+
+function convertToLinesOfEdits(linesOfDiffs: Diff[][], startLineNumber: number) {
+  return flatMap(linesOfDiffs, (diffs, i) => diffsToEdits(diffs, startLineNumber + i));
+}
+
+function diffBy(diffMethod: DiffMethod, x: string, y: string): [Diff[], Diff[]] {
+  const jsDiffChanges: Change[] = jsDiff[diffMethod](x, y);
+  const diffs: Diff[] = diff.convertChangesToDMP(jsDiffChanges);
+
+  if (diffs.length <= 1) {
+    return [[], []];
+  }
+
+  return groupDiffs(diffs);
+}
+
+function diffChangeBlock(
+  changes: ChangeData[],
+  diffMethod: DiffMethod
+): [RangeTokenNode[], RangeTokenNode[]] {
+  /* Convert ChangeData array to two strings representing old source and new source of a change block, like
+  
+  "created_at": "2023-11-20T16:47:52.801Z",
+  "created_by": "elastic",
+  ...
+  
+  and
+
+  "created_at": "1970-01-01T00:00:00.000Z",
+  "created_by": "",
+  ...
+  */
+  const [oldSource, newSource] = changes.reduce(
+    // eslint-disable-next-line @typescript-eslint/no-shadow
+    ([oldSource, newSource], change) =>
+      isDelete(change)
+        ? [oldSource + (oldSource ? '\n' : '') + change.content, newSource]
+        : [oldSource, newSource + (newSource ? '\n' : '') + change.content],
+    ['', '']
+  );
+
+  const [oldDiffs, newDiffs] = diffBy(diffMethod, oldSource, newSource);
+
+  if (oldDiffs.length === 0 && newDiffs.length === 0) {
+    return [[], []];
+  }
+
+  const getLineNumber = (change: ChangeData | undefined) => {
```

Comment:
Can you move this function out of `diffChangeBlock`? It disrupts the reading flow, and doesn't need to be inside

Reply:
Done